### ISSUE TITLE
Expandable/collapsable loop tags bar

### ISSUE THREAD

**Issue by oceanhahn** [aseprite/aseprite#920]
I have a sprite with a few overlapping tags in the animation/layers pane, but if two loop tags begin on the same frame (or even on a nearby frame), the tags can overlap, making the occluded one inaccessible without (as far as I know) deleting the superposed loop. Which I don't want to do.

If we could have some way of either selecting a loop from, maybe a context menu of overlapping ones (generated on-the-spot when multiple tag labels overlap) or have the space loop labels appear in grow and shrink automatically when tags <i>would</i> overlap (example below), that would be dandy!

I made this mock-up of what I was thinking, if it's of any help:
![aseprite tabs](https://cloud.githubusercontent.com/assets/11544843/12056338/b5eb025e-af35-11e5-8659-05979015a8c9.png)

The upper section is the existing form; the lower part is the same data presented in a format I'd imagined based on the interface as it is now. There's also a new button there, that collapses the tags area into a single row (as it currently exists). Perhaps when labels overlap, only the first letter is shown, and the label is offset according to length (with the labels for longer loops occurring last, so that they're above at least some of their frames)? 
It might also be convenient if the loop selection brackets were coloured after the loop labels!

Thanks for considering it!

**Comment by dacap** [comment]
Related to #768

**Comment by jpnk** [comment]
@oceanhahn i'm not sure about collapsing the tag area - tags/animations are of the same importance as layers. they are the third dimension, after layers and frames.
maybe if the tag area grows too big let the user collapse it into a selector (a dropdown list?). i need to reflect on this some more.

**Comment by oceanhahn** [comment]
The basic purpose I'd thought for collapsing the tag/loop area is so that, if the total height of the tags became too deep (that is, it intruded too greatly upon the editor space), the user could press a button to roll them up (and, later, press it again to restore them). Basically, it just hustles them out of the way when they aren't needed and brings them back with a single click, in much the same way that [Tab] hides and reveals the entire pane.

I'm not super-sure how useful it might be considering that, to be meaningfully cumbersome, there would probably have be be at least three stacked sets of layers (fewer, I suppose, if ASEprite is running in a smallish window or on a small or low-res screen), but it might be nice to be able to switch between a full profile of loop labels and a compressed one during viewing.

**Comment by jpnk** [comment]
that's an interesting idea about [Tab]. \<Modifier>+[Tab] could be used to toggle the state of the tag area.

when overlapping tags are collapsed into a single row they look really messy when not colorcoded. what makes it worse is that you can't click on an overlapped tag to bring it to the foreground.
maybe when two overlapping tags start on the same frame, when you click on the foreground one the tag area is temporarily expanded to let you select the right tag. also such tags that cover other tags could be marked somehow e.g. the shadowed tags could stick out from behind them by a couple of pixels.

**Comment by oceanhahn** [comment]
Mm. That was my issue to begin with, actually. I had noted the pulldown menu, but the shadow is a really good addition. :D

**Comment by jpnk** [comment]
this is how tags could work wrt clicking (disregarding overlapping issues):
- when you left-click a tag it is marked as *current*. this is roughly equivalent to selecting its frames and pressing \<F2>. if you left-click it again the mark is removed.
- when you right-click a tag a dialog with its properties is brought up.

when a tag is selected as current you can delete it with \<Delete>.

**Comment by dacap** [comment]
I was thinking about this issue, and I came with this possible solution:

* Moving the mouse over the tags they should expand and you can click one tag from the list to make it active/current
* Right-clicking a tag should popup a menu with options like Properties / Remove / etc

![new-tags](https://cloud.githubusercontent.com/assets/39654/15659424/c827fb42-26a0-11e6-83c5-8f8977323a6a.gif)

**Comment by SolarLune** [comment]
Also interested in this.

@dacap Your solution looks pretty simple! I might recommend making the shadowed tag have brackets pointing to the frame like the foreground one, just darkened according to the color of the tag. 

Otherwise, it seems fine to me!

**Comment by oceanhahn** [comment]
@dacap I really like that idea, too. :D

**Comment by Tranberry** [comment]
what is the UX benefit from having the _a_ animated upward?

**Comment by dacap** [comment]
@Tranberry yesterday I was seeing the animation again and thought _"this should not work in this way"_: the tag that is being focused by the mouse should stay in the same place, and a popup with the other tags should appear above. I'll create another mockup soon.

**Comment by oceanhahn** [comment]
I'm with @SolarLune on having the other tags' range brackets visible, but shaded.

**Comment by 2dchaos** [comment]
I like @oceanhahn original idea, I like having all my tags visible at the same time. There are tags that I use to mark down some animation loop(for example "Idle", "Run") and others that I use to mark down important keyframes on my animation ("Contact", "Passing"), so an option to not loop some tagged animations would be nice.
Also, adding shortcuts to switch between the previows and next tags in a level would be nice.

**Comment by dacap** [comment]
I'm working on this issue, here some screenshots:

![screen shot 2017-03-26 at 9 46 41 pm](https://cloud.githubusercontent.com/assets/39654/24336972/c21e7154-126d-11e7-8576-6b804a4b4769.png)

![screen shot 2017-03-26 at 9 46 46 pm](https://cloud.githubusercontent.com/assets/39654/24336973/c52a54c6-126d-11e7-9616-0ef685c547fd.png)

I think the @oceanhahn idea about collapse/expand the view might be useful (e.g. to hide all overlapped text labels)

**Comment by oceanhahn** [comment]
@dacap Really liking what I'm seeing so far! :D

**Comment by dacap** [comment]
@oceanhahn yeah, I'm trying to color the brackets too (some difficulties here)

**Comment by oceanhahn** [comment]
@dacap Out of curiosity, how are you thinking of going about recolouring the lines? Palette swap?

**Comment by dacap** [comment]
@oceanhahn I can use the `she::Surface::drawColoredRgbaSurface` function to draw a "theme/skin part" with a specific color (black is converted to the color, and alpha channel is kept), the UI uses RGBA (not indexed color). There are two implementations of `drawColoredRgbaSurface`:
* A generic (slow) one: https://github.com/aseprite/aseprite/blob/47f3d540b7a2d961cd13193f4ad2019739adbb86/src/she/common/generic_surface.h#L58
* And the Skia one with a color filter: https://github.com/aseprite/aseprite/blob/27bc151a37d083b5e608d3eddc2f2aaca43d92c2/src/she/skia/skia_surface.h#L430

**Comment by dacap** [comment]
![screen shot 2017-03-27 at 11 36 01 am](https://cloud.githubusercontent.com/assets/39654/24361896/950de2ca-12e1-11e7-9d25-dea1255dc521.png)

**Comment by oceanhahn** [comment]
@dacap Very cool; I'm absolutely looking forward to seeing this in action! Thanks for sharing the specifics, too! :D

**Comment by SolarLune** [comment]
Whoa, cool, that's way better already.

**Comment by oceanhahn** [comment]
I still think the popup-tags idea is a good one too, perhaps when the tags are in a collapsed state. I'm not sure how it would work for tags that overlap _but have offset starting and ending frames,_ but I'm still excited to see how it'll all come together.

Thanks for investing so much effort in this! :D

**Comment by oceanhahn** [comment]
Oh man, is it going to be in 1.1.14? :grinning:

**Comment by dacap** [comment]
@oceanhahn there will not be v1.1.x versions because the code was too different to the beta version 😞   the next version will be v1.2-beta8 (available through Humble Bundle, Gumroad and itch.io too)

**Comment by oceanhahn** [comment]
Aa, I see, so the next updates in the stable branch won't be until v1.2?

**Comment by dacap** [comment]
I think so, anyway after v1.2-beta8 I'd like to fix as many bugs as possible to make v1.2 stable ASAP.

**Comment by oceanhahn** [comment]
Cool-- really looking forward to it! :D

**Comment by fluffbirb** [comment]
v1.2-beta12
I've been using nested tags a lot! Labelling individual frames, loops, and current WIP section. But it seems like every time I erase or draw (perhaps other actions too), it uncollapses all the tags. 

![image](https://user-images.githubusercontent.com/32004169/30507902-38ddd62c-9a59-11e7-8c22-6d81a176882a.png)
https://imgur.com/a/2FiMr
So if I am drawing frames in the WIP tag, then tabbing through my frames with <> so that I don't flip out of the tag, every time I make a mark, all the tags uncollapse, and when I tab frames to D or R, I can't keep flipping, because I'm restricted to the tag D, which is only one frame. 
So I find myself constantly in a loop of 
- collapse to WIP tag
- flip frames to figure out movement
- draw new frame
- tags uncollapse
- start flipping to see if new drawing works
- getting stuck inside a tag other than WIP
- repeat

**Comment by dacap** [comment]
I'm reopening this issue to keep track of the original collapse/expand tags idea. Also requested here: https://community.aseprite.org/t/request-collapse-tags/3349 and in other several places.

**Comment by SpookySquidGames** [comment]
Is there an open issue / still plans for collapsing individual tags horizontally the way you can for layers with the file folders? This issue #920 seems to be dedicated to collapsing tags vertically which is a very different feature (at least on the user end). 

I looked for an open issue to subscribe to for horizontal tag collapsing but it looks like they've all been closed eg https://github.com/aseprite/aseprite/issues/4848(https://github.com/aseprite/aseprite/issues/4848) mentioned above was closed, but linked here? 

Is this issue now covering both types of tag collapsing, both horizontal and vertical?  Really looking forward to horizontal tag collapsing being added since I've got some very long timelines to wrangle. Thanks!
